### ISSUE TITLE
Loss Doesn't  Decrease or Decrease Very Slow

### ISSUE THREAD

**Issue by JialeTao** [NVIDIA/apex#518]
Hi, I have used apex.amp for mix precision model, and choose setting "O1". While everything goes normally except the loss does't decrease during training. The infrence speed get faster and the menmory usage decreased normally. Do you have any suggestion?

**Comment by ptrblck** [comment]
Hi @JialeTao,

is the loss decreasing faster without `amp`?
If not, the issue might be related to your general training procedure, your model, hyperparameters etc.

**Comment by JialeTao** [comment]
> Hi @JialeTao,
> 
> is the loss decreasing faster without `amp`?
> If not, the issue might be related to your general training procedure, your model, hyperparameters etc.

Yeah, when I don't use amp, it decrease faster. I'm confused of this. And my training task is based on image semantic segmentation with input size 1024*512.  CrossEntropy loss is computed on each pixel.

**Comment by ptrblck** [comment]
That sounds weird. Do you have a reproducible code snippet so that we could have a look?

**Comment by JosephDiPalma** [comment]
I had this same issue when using the code from the PyTorch "Training a Classifier" tutorial [https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html](url).  The code I used is:
```python
import torch
import torchvision
import torchvision.transforms as transforms
import torch.optim as optim
from apex import amp
import matplotlib.pyplot as plt
import numpy as np
import torch.nn as nn
import torch.nn.functional as F

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

OPT_LEVEL = "O1"

transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR10(root="./data", train=True, download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=64, shuffle=True, num_workers=2)

classes = ("plane", "car", "bird", "cat", "deer", "dog", "frog", "horse", "ship", "truck")

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 1024, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(1024, 128, 5)
        self.fc1 = nn.Linear(128 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 128 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

net = Net()
net.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=0.1, momentum=0.9)

model, optimizer = amp.initialize(models=net, optimizers=optimizer, opt_level=OPT_LEVEL, verbosity=1)

training_losses = []

for epoch in range(10):  # loop over the dataset multiple times

    for i, data in enumerate(trainloader):
        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data[0].to(device), data[1].to(device)

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        if i % 10 == 9:
            training_losses.append(loss.item())

plt.figure(figsize=(20, 10))
plt.title("Training Loss")
plt.plot(training_losses)
plt.xlabel("Iterations")
plt.ylabel("Loss")
plt.show()
```

I'm running this code on the 19.09 PyTorch Docker container from NVIDIA GPU Cloud.

**Comment by AshStuff** [comment]
@ptrblck  I do have find the same issue. When I use `amp` I can see that the loss does not decrease neither increase but when I don't use `amp` I can see the loss decreasing. I have using `Jensen-shanon divergence loss`. 

```
def entropy(v):
    """
        Entropy loss for probabilistic prediction vectors
        input: batch_size x channels x h x w
        output: batch_size x 1 x h x w
    """
    assert v.dim() == 4
    n, c, h, w = v.size()
    return -torch.sum(torch.mul(v, torch.log2(v + 1e-30))) / (n * h * w * np.log2(c))


class JSD(torch.nn.Module):
    def __init__(self):
        super(JSD, self).__init__()

    def forward(self, outputs):
        softmaxs = [torch.softmax(output, 1) for output in outputs]
        M = (1 / len(softmaxs)) * sum(softmaxs)
        return entropy(M) - ((1 / len(softmaxs) * sum([entropy(i) for i in softmaxs])))

```

**Comment by ptrblck** [comment]
@AshStuff 
Could you post a dummy model or what kind of output is expected to work with your loss function?

**Comment by JialeTao** [comment]
> That sounds weird. Do you have a reproducible code snippet so that we could have a look?
I use DeepLab-Res101 and cross-entropy loss in my model. And I found that the problem comes from the DeepLab-Res101 model or the pretrained parameters I used. Because when I use another with same loss, the problem does not happens. But I can't find that what's wrong with the model. A guess is that it may be connected with batch normalization layer.  Here is my code of  DeepLab-Res101 model. Any datasets based on image semantic segmentation can be used for test.  And pretrained parameters is [[http://vllab.ucmerced.edu/ytsai/CVPR18/DeepLab_resnet_pretrained_init-f81d91e8.pth|here]]. Setting "pretrained=True to load the parameters".

# -*- coding: utf-8 -*-
import torch
import torch.nn as nn
import torch.nn.functional as F

affine_par = True

class Bottleneck(nn.Module):
    expansion = 4

    def __init__(self, inplanes, planes, stride=1, dilation=1, downsample=None, bn_momentum=0.1):
        super(Bottleneck, self).__init__()
        self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, stride=stride, bias=False)  # change
        self.bn1 = nn.BatchNorm2d(planes, affine=affine_par)

        padding = dilation
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=1,  # change
                               padding=padding, bias=False, dilation=dilation)
        self.bn2 = nn.BatchNorm2d(planes, affine=affine_par)

        self.conv3 = nn.Conv2d(planes, planes * 4, kernel_size=1, bias=False)
        self.bn3 = nn.BatchNorm2d(planes * 4, affine=affine_par)

        self.relu = nn.ReLU(inplace=True)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        residual = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)

        out = self.conv3(out)
        out = self.bn3(out)

        if self.downsample is not None:
            residual = self.downsample(x)

        out += residual
        out = self.relu(out)

        return out


class Classifier_Module(nn.Module):
    def __init__(self, inplanes, dilation_series, padding_series, num_classes):
        super(Classifier_Module, self).__init__()
        self.conv2d_list = nn.ModuleList()
        for dilation, padding in zip(dilation_series, padding_series):
            self.conv2d_list.append(
                nn.Conv2d(inplanes, num_classes, kernel_size=3, stride=1, padding=padding, dilation=dilation, bias=True))

        for m in self.conv2d_list:
            m.weight.data.normal_(0, 0.01)

    def forward(self, x):
        out = self.conv2d_list[0](x)
        for i in range(len(self.conv2d_list) - 1):
            out += self.conv2d_list[i + 1](x)
            return out


class ResNetMulti(nn.Module):
    def __init__(self, block, layers, num_classes):
        self.inplanes = 64
        super(ResNetMulti, self).__init__()
        self.conv1 = nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3,
                               bias=False)
        self.bn1 = nn.BatchNorm2d(64, affine=affine_par)
        for i in self.bn1.parameters():
            i.requires_grad = False
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1, ceil_mode=True)  # change
        self.layer1 = self._make_layer(block, 64, layers[0])
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2)
        self.layer3 = self._make_layer(block, 256, layers[2], stride=1, dilation=2)
        self.layer4 = self._make_layer(block, 512, layers[3], stride=1, dilation=4)
        self.layer5 = self._make_pred_layer(Classifier_Module, 1024, [6, 12, 18, 24], [6, 12, 18, 24], num_classes)
        self.layer6 = self._make_pred_layer(Classifier_Module, 2048, [6, 12, 18, 24], [6, 12, 18, 24], num_classes)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
                m.weight.data.normal_(0, 0.01)
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1)
                m.bias.data.zero_()

    def _make_layer(self, block, planes, blocks, stride=1, dilation=1):
        downsample = None
        if stride != 1 or self.inplanes != planes * block.expansion or dilation == 2 or dilation == 4:
            downsample = nn.Sequential(
                nn.Conv2d(self.inplanes, planes * block.expansion,
                          kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(planes * block.expansion, affine=affine_par))
        layers = []
        layers.append(block(self.inplanes, planes, stride, dilation=dilation, downsample=downsample))
        self.inplanes = planes * block.expansion
        for i in range(1, blocks):
            layers.append(block(self.inplanes, planes, dilation=dilation))

        return nn.Sequential(*layers)

    def _make_pred_layer(self, block, inplanes, dilation_series, padding_series, num_classes):
        return block(inplanes, dilation_series, padding_series, num_classes)

    def forward(self, x):
        input_size = x.size()[2:]
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)
        x = self.layer1(x)
        x = self.layer2(x)

        x = self.layer3(x)
        x1 = self.layer5(x)
        # x1 = F.interpolate(x1, size=input_size, mode='bilinear', align_corners=True)

        x2 = self.layer4(x)
        x2 = self.layer6(x2)
        # x2 = F.interpolate(x2, size=input_size, mode='bilinear', align_corners=True)

        # return x2, x1 # changed!
        return x2 # changed!

    def get_1x_lr_params_NOscale(self):
        """
        This generator returns all the parameters of the net except for
        the last classification layer. Note that for each batchnorm layer,
        requires_grad is set to False in deeplab_resnet.py, therefore this function does not return
        any batchnorm parameter
        """
        b = []

        b.append(self.conv1)
        b.append(self.bn1)
        b.append(self.layer1)
        b.append(self.layer2)
        b.append(self.layer3)
        b.append(self.layer4)

        for i in range(len(b)):
            for j in b[i].modules():
                jj = 0
                for k in j.parameters():
                    jj += 1
                    if k.requires_grad:
                        yield k

    def get_10x_lr_params(self):
        """
        This generator returns all the parameters for the last layer of the net,
        which does the classification of pixel into classes
        """
        b = []
        b.append(self.layer5.parameters())
        b.append(self.layer6.parameters())

        for j in range(len(b)):
            for i in b[j]:
                yield i

    def optim_parameters(self, args):
        return [{'params': self.get_1x_lr_params_NOscale(), 'lr': args.base_lr},
                {'params': self.get_10x_lr_params(), 'lr': 10 * args.base_lr}]


def DeeplabMulti(num_classes=21, pretrained=True):
    model = ResNetMulti(Bottleneck, [3, 4, 23, 3], num_classes)

    if pretrained:
        restore_from = './pretrained_model/DeepLab_resnet_pretrained_init-f81d91e8.pth'
        saved_state_dict = torch.load(restore_from)

        new_params = model.state_dict().copy()
        for i in saved_state_dict:
            i_parts = i.split('.')
            if not i_parts[1] == 'layer5':
                new_params['.'.join(i_parts[1:])] = saved_state_dict[i]
        model.load_state_dict(new_params)
    return model

**Comment by ptrblck** [comment]
@JialeTao Thanks for the model definition.
I tried to reproduce the issue using this dummy code snippet:
```python
model = DeeplabMulti(pretrained=False)
model.load_state_dict(torch.load('tmp.pth'))

torch.manual_seed(2809)
data = torch.randn(1, 3, 224, 224)
target = torch.randint(0, 21, (1, 29, 29))

device = 'cuda'
model.to(device)
data = data.to(device)
target = target.to(device)

from apex import amp
use_apex = True

optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)
if use_apex:
    model, optimizer = amp.initialize(model, optimizer, opt_level='O1')
criterion = nn.CrossEntropyLoss()

for idx in range(50):
    optimizer.zero_grad()
    output = model(data)
    loss = criterion(output, target)
    if use_apex:
        with amp.scale_loss(loss, optimizer) as scaled_loss:
            scaled_loss.backward()
    else:
        loss.backward()
    optimizer.step()
    print('iter {}, loss {}'.format(idx, loss.item()))
```
and get similar loss outputs for the FP32 and amp runs:

FP32:
```python
iter 0, loss 5.824324607849121                                                                                                                 
iter 1, loss 4.859009742736816                                                                                                                 
iter 2, loss 4.5960774421691895                                                                                                                
iter 3, loss 4.529462814331055                                                                                                                 
iter 4, loss 4.429961204528809                                                                                                                 
iter 5, loss 4.527955532073975                                                                                                                 
iter 6, loss 4.340815544128418                                                                                                                 
iter 7, loss 4.311011791229248                                                                                                                 
iter 8, loss 4.280165672302246                                                                                                                 
iter 9, loss 3.9953267574310303                                                                                                                
iter 10, loss 3.971621036529541
iter 11, loss 3.8643667697906494
iter 12, loss 3.639624834060669
iter 13, loss 3.5902035236358643
iter 14, loss 3.432408571243286
iter 15, loss 3.2928764820098877
```

AMP O1 (removed first iterations, where the loss scaler was adapted):
```python
iter 9, loss 5.754495620727539                                                                                                                 
iter 10, loss 4.8069257736206055
iter 11, loss 4.522213459014893
iter 12, loss 4.478023529052734
iter 13, loss 4.422145843505859
iter 14, loss 4.348355293273926
iter 15, loss 4.111382007598877
iter 16, loss 3.982872486114502
iter 17, loss 3.9009220600128174
iter 18, loss 3.7114951610565186
iter 19, loss 3.5546793937683105
iter 20, loss 3.146636962890625
iter 21, loss 2.854262351989746
iter 22, loss 2.6023597717285156
iter 23, loss 2.4986727237701416
iter 24, loss 2.118010997772217
```
In fact the loss decreases a bit faster using AMP, but I cannot reproduce a stagnation in the loss decrease.
